src/space.cpp

```diff
@@ -435,9 +340,128 @@ void makeNaCl(Space &space, int num_particles, const Geometry::Chameleon &geomet
+            spc.groups.back().resize(0);
+        }
+    }
+}
+bool InsertMoleculesInSpace::setPositionsForTrailingGroups(Space &spc, int num_molecules,
+                                                           const Faunus::ParticleVector &particles,
+                                                           const Point &offset) {
+    assert(spc.groups.size() >= num_molecules);
+    if (particles.size() == num_molecules * (spc.groups.end() - num_molecules)->traits().atoms.size()) {
+        size_t cnt = spc.p.size() - particles.size();
+        for (auto &i : particles) {
+            auto pos = i.pos + offset;
+            if (spc.geo.collision(pos) == false) {
+                spc.p.at(cnt++).pos = pos;
+            } else {
+                faunus_logger->warn("position outside box");
+                return false;
+            }
+        }
+        if (cnt == particles.size()) {
+            for (auto it = spc.groups.end() - num_molecules; it != spc.groups.end(); ++it) {
+                it->cm = Geometry::massCenter(it->begin(), it->end(), spc.geo.getBoundaryFunc(), -it->begin()->pos);
+            }
+        }
+    } else {
+        faunus_logger->error("wrong number of atoms");
+        return false;
+    }
+    return true;
+}
+void InsertMoleculesInSpace::insertImplicitGroups(const MoleculeData &mol, Space &spc, int num_molecules) {
+    assert(mol.isImplicit());
+    spc.getImplicitReservoir()[mol.id()] = num_molecules;
+}
+void InsertMoleculesInSpace::insertMolecules(const json &json_array, Space &spc) {
+    spc.clear();
+    assert(spc.geo.getVolume() > 0);
+    if (!json_array.is_array()) {
+        throw ConfigurationError("syntax error in insertmolecule");
+    }
+    for (auto &obj : json_array) { // loop over array of molecules
+        if (!obj.is_object() || obj.size() != 1) {
+            throw ConfigurationError("syntax error in insertmolecule");
+        }
+        for (auto &[molname, properties] : obj.items()) {
+            if (auto moldata = findName(Faunus::molecules, molname); moldata != Faunus::molecules.end()) {
+                int num_molecules = 0; // number of groups to insert
+                if (auto it = properties.find("N"); it != properties.end()) {
+                    num_molecules = it->get<int>();
+                } else {
+                    double concentration = properties.at("molarity").get<double>() * 1.0_molar;
+                    num_molecules = std::round(concentration * spc.geo.getVolume());
+                    if (concentration > pc::epsilon_dbl) {
+                        double rel_error = (concentration - num_molecules / spc.geo.getVolume()) / concentration;
+                        if (rel_error > 0.01) {
+                            faunus_logger->warn("{}: initial molarity differs by {}% from target value", molname,
+                                                rel_error * 100);
+                        }
+                    }
+                }
```

Comment:
A function returning `num_molecules` and `rel_error` would be nice. The remaining code here would just log a warning if `rel_error` is to big.

Reply:
next time